File: prototypes/AI_Deltas/LKB/DX7F9L2Q1A3B/src/bias.js
import { clamp } from './math.js';

export class BiasField {
  constructor(size, opts) {
    this.size = size;
    this.n = size ** 3;
    this.decay = opts.decay;
    this.strength = opts.strength;
    this.radius = opts.radius;
    this.bias = new Float32Array(this.n);
  }

  decayField() {
    for (let i = 0; i < this.n; i += 1) {
      this.bias[i] *= this.decay;
    }
  }

  injectSpectrum(left, right) {
    this.decayField();
    const bins = Math.max(left.length, right.length);
    if (!bins) return;

    for (let b = 0; b < bins; b += 1) {
      const aL = left[b] ?? left[left.length - 1] ?? 0;
      const aR = right[b] ?? right[right.length - 1] ?? 0;
      const energy = (aL + aR) * 0.5;
      const pan = aR - aL;
      const z = Math.floor((b / Math.max(1, bins - 1)) * (this.size - 1));
      const x = clamp(Math.floor(this.size / 2 + pan * (this.size * 0.18)), 0, this.size - 1);
      const y = Math.floor(this.size / 2);
      this.addKernel(x, y, z, energy);
    }
  }

  addKernel(cx, cy, cz, energy) {
    const r = this.radius;
    const str = this.strength * energy;
    if (str === 0) return;
    for (let dz = -r; dz <= r; dz += 1) {
      for (let dy = -r; dy <= r; dy += 1) {
        for (let dx = -r; dx <= r; dx += 1) {
          const d2 = (dx * dx + dy * dy + dz * dz) / (r * r || 1);
          if (d2 > 1) continue;
          const kernel = Math.exp(-d2 * 1.8);
          const idx = this.idx(cx + dx, cy + dy, cz + dz);
          if (idx >= 0) this.bias[idx] = clamp(this.bias[idx] + str * kernel, -0.3, 0.3);
        }
      }
    }
  }

  idx(x, y, z) {
    if (x < 0 || y < 0 || z < 0 || x >= this.size || y >= this.size || z >= this.size) return -1;
    return x + y * this.size + z * this.size * this.size;
  }
}

File: prototypes/AI_Deltas/LKB/LKB20250326X7A9/src/lkb_delta/bias.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Iterable, List, Tuple


@dataclass
class BiasField:
    size: int
    bin_count: int
    decay: float
    strength: float
    radius: int

    def __post_init__(self) -> None:
        self.bias: List[List[List[float]]] = [
            [[0.0 for _ in range(self.size)] for _ in range(self.size)]
            for _ in range(self.size)
        ]

    @property
    def amplitude(self) -> float:
        total = sum(abs(v) for v in self._iter_values())
        count = self.size ** 3
        return total / count if count else 0.0

    def _iter_values(self) -> Iterable[float]:
        for plane in self.bias:
            for row in plane:
                for value in row:
                    yield value

    def flatten(self) -> List[float]:
        return list(self._iter_values())

    def decay_bias(self) -> None:
        for x in range(self.size):
            for y in range(self.size):
                for z in range(self.size):
                    self.bias[x][y][z] *= self.decay

    def _inject_single(self, center: Tuple[int, int, int], power: float) -> None:
        cx, cy, cz = center
        r = self.radius
        for x in range(max(cx - r, 0), min(cx + r + 1, self.size)):
            for y in range(max(cy - r, 0), min(cy + r + 1, self.size)):
                for z in range(max(cz - r, 0), min(cz + r + 1, self.size)):
                    dx, dy, dz = x - cx, y - cy, z - cz
                    dist2 = dx * dx + dy * dy + dz * dz + 1e-6
                    falloff = math.exp(-dist2 / max(r * r, 1))
                    self.bias[x][y][z] += power * falloff

    def inject_from_bins(self, left: List[float], right: List[float]) -> None:
        """Map stereo bins into the lattice.

        - Low freq => back (higher z), high freq => front (lower z).
        - Pan => x-axis tilt, magnitude => injection power.
        - Centered on midplane y to keep influence gentle (SV3/SV9).
        """

        if len(left) != self.bin_count or len(right) != self.bin_count:
            raise ValueError("bin arrays must match configured bin_count")

        mid_y = self.size // 2
        for idx, (l_val, r_val) in enumerate(zip(left, right)):
            power = (abs(l_val) + abs(r_val)) * 0.5
            if power <= 0:
                continue

            pan_den = abs(l_val) + abs(r_val) + 1e-6
            pan = (r_val - l_val) / pan_den
            x_pos = int(min(max((pan + 1) * 0.5 * (self.size - 1), 0), self.size - 1))
            z_pos = int(min(max((1 - idx / max(self.bin_count - 1, 1)) * (self.size - 1), 0), self.size - 1))
            center = (x_pos, mid_y, z_pos)
            self._inject_single(center, power * self.strength)
